src/libstd/sys/unix/ext/net/ancillary.rs

```diff
@@ -0,0 +1,615 @@
+                match (*cmsg).cmsg_type {
+                    libc::SCM_RIGHTS => AncillaryData::as_rights(data),
+                    #[cfg(any(
+                        target_os = "linux",
+                        target_os = "android",
+                        target_os = "emscripten",
+                        target_os = "fuchsia",
+                        target_env = "uclibc",
+                    ))]
+                    libc::SCM_CREDENTIALS => AncillaryData::as_credentials(data),
+                    #[cfg(any(
+                        target_os = "netbsd",
+                        target_os = "openbsd",
+                        target_os = "freebsd",
+                        target_os = "dragonfly",
+                        target_os = "macos",
+                        target_os = "ios",
+                    ))]
+                    libc::SCM_CREDS => AncillaryData::as_credentials(data),
+                    _ => panic!("Unknown cmsg type"),
+                }
+            } else {
+                panic!("Unknown cmsg level");
+            }
+        }
+    }
+}
+
+#[cfg(any(
+    target_os = "haiku",
+    target_os = "solaris",
+    target_os = "illumos",
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "freebsd",
+    target_os = "dragonfly",
+    target_os = "openbsd",
+    target_os = "netbsd",
+    target_os = "linux",
+    target_os = "android",
+    target_os = "emscripten",
+    target_os = "fuchsia",
+    target_env = "uclibc",
+))]
+#[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+pub struct Messages<'a> {
+    buffer: &'a [u8],
+    current: Option<&'a libc::cmsghdr>,
+}
+
+#[cfg(any(
+    target_os = "haiku",
+    target_os = "solaris",
+    target_os = "illumos",
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "freebsd",
+    target_os = "dragonfly",
+    target_os = "openbsd",
+    target_os = "netbsd",
+    target_os = "linux",
+    target_os = "android",
+    target_os = "emscripten",
+    target_os = "fuchsia",
+    target_env = "uclibc",
+))]
+#[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+impl<'a> Iterator for Messages<'a> {
+    type Item = AncillaryData<'a>;
+
+    fn next(&mut self) -> Option<AncillaryData<'a>> {
+        unsafe {
+            let msg = libc::msghdr {
+                msg_name: null_mut(),
+                msg_namelen: 0,
+                msg_iov: null_mut(),
+                msg_iovlen: 0,
+                msg_control: self.buffer.as_ptr() as *mut _,
+                msg_controllen: self.buffer.len(),
+                msg_flags: 0,
+            };
+
+            let cmsg = if let Some(current) = self.current {
+                libc::CMSG_NXTHDR(&msg, current)
+            } else {
+                libc::CMSG_FIRSTHDR(&msg)
+            };
+
+            let cmsg = cmsg.as_ref()?;
+            self.current = Some(cmsg);
+            let ancillary_data = AncillaryData::from(cmsg);
```

Comment:
This should propagate the error, and the iterator can iterate over `Result<AncillaryData<'a>, AncillaryError>`. `AncillaryError` should be `#[non_exhaustive]`, with variants for `UnknownType(i32)` and `UnknownLevel(i32)`.